File: clients/cpp/model/Building.cpp
#include "Building.hpp"

namespace model {

Building::Building(model::BuildingType buildingType, int health, int workDone, int lastTickTasksDone) : buildingType(buildingType), health(health), workDone(workDone), lastTickTasksDone(lastTickTasksDone) { }

// Read Building from input stream
Building Building::readFrom(InputStream& stream) {
    model::BuildingType buildingType = readBuildingType(stream);
    int health = stream.readInt();
    int workDone = stream.readInt();
    int lastTickTasksDone = stream.readInt();
    return Building(buildingType, health, workDone, lastTickTasksDone);
}

// Write Building to output stream
void Building::writeTo(OutputStream& stream) const {
    stream.write((int)(buildingType));
    stream.write(health);
    stream.write(workDone);
    stream.write(lastTickTasksDone);
}

// Get string representation of Building
std::string Building::toString() const {
    std::stringstream ss;
    ss << "Building { ";
    ss << "buildingType: ";
    ss << buildingTypeToString(buildingType);
    ss << ", ";
    ss << "health: ";
    ss << health;
    ss << ", ";
    ss << "workDone: ";
    ss << workDone;
    ss << ", ";
    ss << "lastTickTasksDone: ";
    ss << lastTickTasksDone;
    ss << " }";
    return ss.str();
}

bool Building::operator ==(const Building& other) const {
    return buildingType == other.buildingType && health == other.health && workDone == other.workDone && lastTickTasksDone == other.lastTickTasksDone;
}

}

size_t std::hash<model::Building>::operator ()(const model::Building& value) const {
    size_t result = 0;
    result ^= std::hash<model::BuildingType>{}(value.buildingType) + 0x9e3779b9 + (result << 6) + (result >> 2);
    result ^= std::hash<int>{}(value.health) + 0x9e3779b9 + (result << 6) + (result >> 2);
    result ^= std::hash<int>{}(value.workDone) + 0x9e3779b9 + (result << 6) + (result >> 2);
    result ^= std::hash<int>{}(value.lastTickTasksDone) + 0x9e3779b9 + (result << 6) + (result >> 2);
    return result;
}

File: clients/rust/src/model/building.rs
use super::*;

/// A building
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Building {
    /// Building's type
    pub building_type: model::BuildingType,
    /// Current health
    pub health: i32,
    /// Amount of work done for current task
    pub work_done: i32,
}

impl trans::Trans for Building {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.building_type.write_to(writer)?;
        self.health.write_to(writer)?;
        self.work_done.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let building_type: model::BuildingType = trans::Trans::read_from(reader)?;
        let health: i32 = trans::Trans::read_from(reader)?;
        let work_done: i32 = trans::Trans::read_from(reader)?;
        Ok(Self {
            building_type,
            health,
            work_done,
        })
    }
}